### PR TITLE
Styled Theming + media template introduction

### PR DESCRIPTION
**Pull request by rossbulat**
* Introduces `styled-theming`, installed in `ui-app`. `substrate` and `polkadot` themes as per `settings.uiTheme`
* `<ThemeProvider />` wraps everything in `Apps.tsx`
* media templates for `MOBILE`, `TABLET` and `DESKTOP` imported from `ui-app/src/media.ts` Media templates are "mobile first" adopting a `min-width` approach to screen sizing, following the conventions used in `media.css`.
* Converted the menu toggle media queries to styled theming

Following @jacogr guidelines to introduce `styled`  as and when files are touched.

Demonstration of embedded theme config:
```
const Toggle = styled.img`
  background: ${theme('theme', {
    substrate: '#333',
    polkadot: 'none'
  })};
  padding: ${theme('theme', {
    substrate: '4px',
    polkadot: 'none'
  })};
...
```

Demonstation of media template usage:
```
const Toggle = styled.img`
   ...
  ${media.TABLET`
    opacity: 0 !important;
    top: -2.9rem !important;
  `}
`;

```

### CODE REVIEW

File: packages/apps/src/SideBar/index.tsx [hunk 1]
```diff
@@ -27,6 +30,38 @@ type Props = I18nProps & {
   toggleMenu: () => void
 };
 
+const Toggle = styled.img`
+  background: ${theme('theme', {
```
Comment (jacogr) [inline packages/apps/src/SideBar/index.tsx]: My only comment here is that I almost feel that these values (i.e. background), should be set somewhere globally - it makes it easier to align colors as well. (That is the nice things about the current theme location - it is in a single place in terms of global colors.)

Just worried that we end up repeating some stuff spread-out all-over?

Thoughts?

File: packages/ui-app/src/media.ts [hunk 2]
```diff
@@ -5,13 +5,15 @@
...
 
-export const media = Object
+export const media: any = Object
  .keys(ScreenSizes)
- .reduce((acc, label) => {
-   acc[label] = (...args) => css`
-     @media (min-width: ${ScreenSizes[label] / 16}em) {
-      ${css(...args)}
+ .reduce((acc: any, label: any) => {
+   let size: any = ScreenSizes[label];
```
[the remark this answers is not in the document]
Comment (rossbulat) [reply packages/ui-app/src/media.ts]: Good catch !

File: packages/apps/src/SideBar/index.tsx (hunk 1 above)
Comment (rossbulat) [reply packages/apps/src/SideBar/index.tsx]: Yes, perhaps we could utilise refs to refer to abstract `styled.css` objects and inject them elsewhere - will have a look at implementations and get back to this.

**Comment by rossbulat** [comment]
> My only comment here is that I almost feel that these values (i.e. background), should be set somewhere globally - it makes it easier to align colors as well. (That is the nice things about the current theme location - it is in a single place in terms of global colors.)

I have abstracted base colours in `ui-app/src/styles/theme.ts` (is there a better place?).

Right now just have the primary color:
```
export const primaryColor = theme('theme', {
  substrate: '#333',
  polkadot: '#E6007A'
});
```

Which is then used in sidebar togggle:

```
import { primaryColor } from '@polkadot/ui-app/styles/theme';


const Toggle = styled.img`
  background: ${theme('theme', {
    substrate: primaryColor,
    polkadot: 'none'
  })};
 ...
```
I'd be happy to build upon this.

**Comment by Stefie** [comment]
@jacogr fromm what I understood in earlier conversations your idea was also to make it easy for outside projects to just fork this repo and create their own version with a custom theme. Did I understand that correctly?

If so, it really makes things complicated when there are theme references in every component like:
```
const Toggle = styled.img`
  padding: ${theme('theme', {
    substrate: '4px',
    polkadot: 'none'
  })}`;
```

That works well for an internal implementation like a `dark` and a `light` theme in a closed source application, but not if you're aiming for something like a whitelabel solution that is easy customizable.
Putting the different theme declarations inside the components means that a person who wants to create their own theme needs to touch every single themed component and add the declarations for their own theme there.

It makes more sense (and is common practice) to put all themeable style declaration into one file/ folder that functions as a theme and then inject that theme to the `ThemeProvider` : It's also the recommended way in the styled components docs. https://www.styled-components.com/docs/advanced#theming.

 It's also how it's done in a very slim version in  `substrate-light-ui` https://github.com/paritytech/substrate-light-ui/blob/500bfd37edfb116bbd3a506a067cb5e00e6a610b/packages/ui-components/src/globalStyle.tsx#L27

That way people who want to add their own theme only need to touch one file/ folder and have all the customizable declarations in one place and they don't have leftover style declarations in the component code that they're never using .

If a whitelabel-like solution is not the goal, forget everything I said above.

**Comment by rossbulat** [comment]
While possible there are some [issues](https://jamie.build/styled-theming.html) with this approach. 

I’d lean towards abstracting ‘substrate’ and ‘polkadot’ themes and treat them as ‘default’ and ‘custom’. User could then fork the project and customise ‘theme.ts’ with their liking.

**Comment by jacogr** [comment]
Bear in mind that we have very limited resources on here, so while I appreciate the route other projects take and/or may take now or in the future, it doesn't affect this repo.

I am not 100% with the defs inside the components, but also need to actually see all this play out and in-action and then determine the actual best route based on the dangling bits. I only know that this will end up looking very different 2 months from now, but prefer code in the repo above everything all-else - so it can be in-hand, visible and active instead of just floating in the air.

As always, PRs are welcome where things can be improved.

**Comment by Stefie** [comment]
The link you posted doesn't really address the problem of making the themes easily customizable. As I said above, for a closed application with different themes/ modes this solution sounds good, but in my experience (and also the creators of many UI libraries) having everything in one place is exactly what you need to make theming easy and effortless.

The approach of having some declarations in the `theme.ts` and additional declarations in the component would still leave the user with the need to change them in two places. 

And colors are one part of a theme, paddings, margins, fonts etc yet another. You just shouldn't need to touch single components in order to change a theme.

**Comment by Stefie** [comment]
@jacogr I'm just afraid that we're actually making it harder for people to customise their fork rather than easier. That's why I was asking where the focus is, whether it's internal theming or having a solution that is easy to whitelabel

**Comment by rossbulat** [comment]
Agreed with your points, I would also keep the customisable theming in `theme.ts`.

The menu toggle was a bit of an edge case as it involved the 2 layer abstraction because of the padding of the Substrate logo. I guess this could be addressed quite simply. But on the other side of the coin demonstrates the flexibility of styled theming.

**Comment by jacogr** [comment]
First and foremost - 

- There is not enough people looking at apps to solve the world's problems. Case in point, the only UI work that has recently gone it that is non-functionality related is by an external contributor.  (And a lot of this work has been focussed on apps itself on any kind of device)

- The primary focus is on making apps great, then looking at anything else. And for this, need concrete use-cases instead of just dreams of what works and what doesn't. (Unlike repos like the API where the primary focus is a toolbox, here the primary focus is the end-user of apps - clicking)

Would be happy to consider any concrete PRs as a proposal for improvement, but the days are too short to talk around things, instead of getting code into the repo.

**Pull request merged**